### PR TITLE
Clean up http2 interop tests

### PR DESCRIPTION
**Pull request by carl-mastrangelo**
Handles Goaway frame, and http1 responses (which some servers send)

### CODE REVIEW

**Comment by jtattermusch** [comment]
I'll be gone for a week, so reassigning to @adelez. Thanks!

**Comment by carl-mastrangelo** [comment]
+cc: @ctiller

File: tools/http2_interop/http2interop.go [hunk 1]
```diff
@@ -279,16 +286,44 @@ func testTLSBadCipherSuites(ctx *HTTP2InteropCtx) error {
...
-			return err
+			return nil, err
 		}
-		if gf, ok := f.(*GoAwayFrame); ok {
-			return fmt.Errorf("Got goaway frame %d", gf.Code)
+		if gf, ok := f.(*GoAwayFrame); !ok {
+			continue
+		} else {
+			return gf, nil
 		}
```
Comment (adelez) [inline tools/http2_interop/http2interop.go]: Not familiar with go but do you need to wait a little before the next iteration?

File: tools/http2_interop/http2interop.go [hunk 2]
```diff
@@ -195,21 +191,22 @@ func testUnknownFrameType(ctx *HTTP2InteropCtx) error {
...
+	conn.SetDeadline(time.Now().Add(defaultTimeout))
 
-	if _, err := c.Write([]byte(prefacePrefix)); err != nil {
+	if _, err := conn.Write([]byte(prefacePrefix)); err != nil {
 		return err
 	}
 
-	buf := make([]byte, 256)
-	for ; err == nil; _, err = c.Read(buf) {
+	if _, err := expectGoAwaySoon(conn); err != nil {
```
Comment (adelez) [inline tools/http2_interop/http2interop.go]: Why not expectGoAway?

File: tools/http2_interop/http2interop.go (hunk 1 above)
Comment (carl-mastrangelo) [inline tools/http2_interop/http2interop.go]: parseFrame() will actually block waiting on a full frame to come over the connection.  Once we get the go away frame it is safe to return.

File: tools/http2_interop/http2interop.go (hunk 2 above)
Comment (carl-mastrangelo) [inline tools/http2_interop/http2interop.go]: expectGoAway is for when the very first frame should be a Go away.  This is for things like bad TLS configuration.   Sometimes it isn't feasible for the TLS terminator to teardown the connection, so the spec allows sending a goaway frame with inadequate security before any other frames.

However, most servers, upon getting a good TLS connection, will immediately send their preface before actually receiving any bytes from the client, in order to speed up the handshake.  In this test, I am intentionally sending a bad preface to the server, which means the server preface may already be on the wire.  This code expects a few valid frames (like settings, continuations, and window updates) before getting a goaway frame from the server.

File: tools/http2_interop/http2interop_test.go [hunk 3]
```diff
@@ -90,13 +99,22 @@ func TestUnknownFrameType(t *testing.T) {
...
+}
+
 func TestTLSApplicationProtocol(t *testing.T) {
 	if *testCase != "tls" {
 		t.SkipNow()
 	}
 	ctx := InteropCtx(t)
 	err := testTLSApplicationProtocol(ctx)
-	matchError(t, err, "EOF")
+	matchError(t, err, "EOF", "broken pipe")
```
Comment (adelez) [inline tools/http2_interop/http2interop_test.go]: Does this print err?

**Comment by adelez** [comment]
LGTM, but since Go is new to me. You may want to wait for LGTM from ctiller.

**Comment by carl-mastrangelo** [comment]
This is blocking me from continuing work, so I will submit it.  Thanks for the review!

**Comment by carl-mastrangelo** [comment]
Test results says no failures, so merging.

**Pull request merged**